utils/shallow_water/linear.py

```diff
@@ -0,0 +1,37 @@
+
+import firedrake as fd
+
+# # # === --- finite element forms --- === # # #
+
+# mass forms for depth and velocity fields
+
+
+def form_mass_h(mesh, h, p):
+    return (p*h)*fd.dx
+
+
+def form_mass_u(mesh, u, w):
+    return fd.inner(u, w)*fd.dx
+
+
+def form_mass(mesh, h, u, p, w):
+    return form_mass_h(mesh, h, p) + form_mass_u(mesh, u, w)
+
+
+# spatial forms for depth and velocity fields
+
+def form_function_h(mesh, H, h, u, p):
+    return (H*p*fd.div(u))*fd.dx
+
+
+def form_function_u(mesh, g, f, h, u, w):
+    outward_normals = fd.CellNormal(mesh)
+
+    def perp(u):
```

Comment:
You're defining perp twice in this file. Maybe better to have perp as another argument? because in a planar geometry, perp(u) = [-u[1], u[0]].

Reply:
I don't see the second definition, can you point me to it?

Happy to put it as a second argument if that's better. The advantage being that you can give it a more efficient version for planar geometry? Or something else?

Reply:
you defined it in the nonlinear equation set as well

Reply:
Seeing as perp was a one-liner, I defaulted the perp argument straight to fd.cross
This means that the passed perp function needs two arguments, but the normals argument can always be ignored.
See commit [372afa1](https://github.com/colinjcotter/asQ/pull/35/commits/372afa16deb4a2f5c1c6387588742972f3519a3d)